demo/lightning/checkpoint/ray/train.py

```diff
@@ -0,0 +1,143 @@
+from ray.train import RunConfig, CheckpointConfig
+from ray.train.lightning import RayLightningEnvironment, RayTrainReportCallback, prepare_trainer
+
+import lightning as pl
+from lightning import Trainer
+import torch
+from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy
+from torch.distributed.fsdp import BackwardPrefetch
+import pandas as pd
+from datasets import load_dataset
+from transformers import AutoTokenizer, AutoModelForCausalLM
+
+from dataflux_pytorch.lightning import DatafluxLightningCheckpoint
+
+class BertTinyModel(pl.LightningModule):
+    def __init__(self, lr=2e-5, eps=1e-8):
+        super().__init__()
+        self.save_hyperparameters()
+        self.lr = lr
+        self.eps = eps
+        self.model = AutoModelForCausalLM.from_pretrained("prajjwal1/bert-tiny")
+
+    def forward(self, batch):
+        outputs = self.model(
+            batch["input_ids"],
+            attention_mask=batch["attention_mask"],
+            labels=batch["labels"]
+        )
+        return outputs.loss
+
+    def training_step(self, batch, batch_idx):
+        loss = self.forward(batch)
+        self.log("train_loss", loss, prog_bar=True, on_step=True)
+        return loss
+
+    def configure_optimizers(self):
+        if self.global_rank == 0:
+            print(self.trainer.model)
+        return torch.optim.AdamW(self.trainer.model.parameters(), lr=self.lr, eps=self.eps)
+
+def split_text(batch: pd.DataFrame) -> pd.DataFrame:
+    text = list(batch["text"])
+    flat_text = "".join(text)
+    split_text = [
+        x.strip()
+        for x in flat_text.split("\n")
+        if x.strip() and not x.strip()[-1] == ":"
+    ]
+    return pd.DataFrame(split_text, columns=["text"])
+
+
+def tokenize(batch: pd.DataFrame) -> dict:
+    tokenizer = AutoTokenizer.from_pretrained("databricks/dolly-v2-7b", padding_side="left")
+    tokenizer.pad_token = tokenizer.eos_token
+    ret = tokenizer(
+        list(batch["text"]),
+        truncation=True,
+        max_length=256,
+        padding="max_length",
+        return_tensors="np",
+    )
+    ret["labels"] = ret["input_ids"].copy()
+    return dict(ret)
+
+def train_func(config):
+    lr = 2e-5
+    eps = 1e-8
+    batch_size_per_worker = config["flags"].batch_size
+    model = BertTinyModel(lr=lr, eps=eps)
+    strategy = config["strategy"]
+    ckpt = DatafluxLightningCheckpoint(project_name=config["flags"].gcp_project,
+                                           bucket_name=config["flags"].gcs_bucket)
+
+    train_ds = ray.train.get_dataset_shard("train")
+    train_dataloader = train_ds.iter_torch_batches(batch_size=batch_size_per_worker)
+
+    trainer = Trainer(
+        default_root_dir=config["flags"].default_root_dir,
+        max_epochs=config["flags"].epochs,
+        devices="auto",
+        max_steps=1,
+        accelerator="auto",
+        strategy=strategy,
+        plugins=[RayLightningEnvironment(),ckpt],
```

Comment:
When I remove ckpt here it doesn't seem to make any difference in behavior, and my checkpoint still shows up in GCS. I also see from the metrics that the writes to my bucket don't have the "dataflux" user-agent. This makes me think that Ray is probably checkpointing via gcsfs by default, and ignoring the checkpointer you passed in here. Were you able to confirm that your checkpoint code was called at all? If not we can still check this in as a good example of how to configure Ray to checkpoint for GCS (and potentially get baseline metrics from it), but I don't think it really uses Dataflux.

Reply:
I did confirm the dataflux checkpointing code was called because without configuration it results in error but Ray still may be able to overwrite some behavior. Let me dig into the stack trace more